autox/informer.go

```diff
@@ -3,154 +3,359 @@ package autox
 // informer.go - informer(s) to watch desired resources/namespaces
 
 import (
+	"bytes"
+	"context"
+	_ "embed"
+	"errors"
 	"fmt"
-	"reflect"
+	"hash/maphash"
+	"sync"
 
+	"github.com/iter8-tools/iter8/base"
 	"github.com/iter8-tools/iter8/base/log"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
 	"k8s.io/client-go/dynamic/dynamicinformer"
 	"k8s.io/client-go/tools/cache"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 const (
-	autoxLabel = "iter8.tools/autox-group"
+	autoXLabel   = "iter8.tools/autox-group"
+	appLabel     = "app.kubernetes.io/name"
+	versionLabel = "app.kubernetes.io/version"
+	trackLabel   = "iter8.tools/track"
 )
 
+var m sync.Mutex
+
+//go:embed application.tpl
+var tplStr string
+
 type chartAction int64
 
 const (
 	releaseAction chartAction = 0
 	deleteAction  chartAction = 1
 )
 
-func getReleaseName(chartGroupName string, chartName string) string {
-	return fmt.Sprintf("autox-%s-%s", chartGroupName, chartName)
+// applicationValues is the values for the application template
+type applicationValues struct {
+	// name is the name of the application
+	name string
+
+	// name is the namespace of the application
+	namespace string
+
+	// owner is the spec group secret for this application
+	owner struct {
+		name string
+		uid  string
+	}
+
+	// chart is the Helm chart for this application
+	chart struct {
+		name    string
+		values  map[string]interface{}
+		version string
+	}
+}
+
+// the name of a release will depend on:
+//
+//	the name of the releaseSpec,
+//	the ID of the spec within the releaseSpec, and
+//	the set of (pruned) labels that triggers this release
+func getReleaseName(group string, releaseSpecID string, prunedLabels map[string]string) string {
+
+	// use labels relevant to autoX to create a random hash value
+	// this value will be appended as a suffix in the release name
+	var hasher maphash.Hash
+	// specGroupName and specID are always hashed
+	_, _ = hasher.WriteString(group)
+	_, _ = hasher.WriteString(releaseSpecID)
+
+	// hash app label
+	app := prunedLabels[appLabel]
+	_, _ = hasher.WriteString(app)
+
+	// hash version label
+	version := prunedLabels[versionLabel]
+	_, _ = hasher.WriteString(version)
+
+	// hash track label
+	track := prunedLabels[trackLabel]
+	_, _ = hasher.WriteString(track)
+
+	nonce := fmt.Sprintf("%05x", hasher.Sum64())
+	nonce = nonce[:5]
+	return fmt.Sprintf("autox-%s-%s-%s", group, releaseSpecID, nonce)
 }
```

Comment:
I presume that we want the same value for nonce for the same input. This implementation creates a new hasher each time it is called resulting in a different seed. The different seed means the return value might be different. To address this either explictily set the seed to a common value or use a single (global) hasher and call hasher.Reset() before each use.

We have/had the same problem in the the abn lookup code.